---
title: "On the Robustness of AlphaFold: A COVID-19 Case Study"
date: 2023-01-10T17:31:39.000Z
author: "Ismail Alkhouri, Sumit Jha, Andre Beckus, George Atia, Alvaro Velasquez, Rickard Ewetz, Arvind Ramanathan, Susmit Jha"
weight: 2
showToc: true
TocOpen: true
draft: false
hidemeta: false
comments: false
description: "Important disclaimer: the following content is AI-generated, please make sure to fact check the presented information by reading the full paper."
disableHLJS: true # to disable highlightjs
disableShare: false
hideSummary: false
searchHidden: false
ShowReadingTime: false
ShowBreadCrumbs: false
ShowPostNavLinks: false
ShowWordCount: true
ShowRssButtonInSectionTermList: false
UseHugoToc: false
cover:
    image: "thumbnails/2301-04093v1.webp" # image path/url
    alt: "On the Robustness of AlphaFold: A COVID-19 Case Study" # alt text
    caption: "The full paper is available [here](https://arxiv.org/abs/2301.04093)." # display caption under cover
    relative: false # when using page bundles set this to true
    hidden: false # only hide on current single page
---

# Link to paper
The full paper is available [here](https://arxiv.org/abs/2301.04093).


# Abstract
- Protein folding neural networks (PFNNs) such as AlphaFold predict remarkably accurate structures of proteins
- However, the robustness of such networks has heretofore not been explored
- This is particularly relevant given the broad social implications of such technologies and the fact that biologically small perturbations in the protein sequence do not generally lead to drastic changes in the protein structure
- In this paper, we demonstrate that AlphaFold does not exhibit such robustness despite its high accuracy
- This raises the challenge of detecting and quantifying the extent to which these predicted protein structures can be trusted
- To measure the robustness of the predicted structures, we utilize (i) the root-mean-square deviation (RMSD) and (ii) the global distance test (GDT) similarity measure
- We prove that the problem of minimally perturbing protein sequences to fool protein folding neural networks is NP-complete

# Paper Content

## Introduction
- Proteins are composed of amino acids
- Naturally occurring proteins are bio-polymers typically composed of 20 amino acids
- Protein folding is a challenge due to the variability of protein sequences
- AlphaFold is a neural network that is successful in predicting protein structures
- Adversarial sequences are sequences that vary only in five residues out of hundreds or thousands of residues and can result in very different protein structures
- The problem of adversarial attacks on PFNN is NP-complete
- Henikoff and Henikoff 1992 is a metric that is used to identify a space of similar protein sequences
- The RMSD (GDT-TS) is a metric that is used to measure the adversarial robustness of a protein sequence

## Summary and Related work
- Protein folding neural networks are usually accurate
- Small changes in the sequence of a protein usually do not lead to drastic changes in the protein structure
- Similar proteins usually produce similar structures
- Manipulating the multiple sequence alignment step of Al-phaFold has been studied in (Stein and Mchaourab 2021)
- However, there, the goal is not to study the robustness of the protein folding neural networks, but rather to enhance the prediction capability of AlphaFold
- In (Jha et al. 2021), the authors show results for a few proteins and do not consider all the standard metrics for measuring the output structures.
- In contrast, in this paper, we present results for more than 100 sequences, derive a complexity proof for the problem of finding adversarial protein sequences, and, based on the CASP competition, utilize all the standard metrics for measuring the output structures.

### Robustness Metric using Adversarial Attacks
- The similar-sequence implies similar-structure paradigm dictates that PFNNs should make robust predictions.
- Given a protein sequence of n residues S = s 1 s 2 . . . s n with a threedimensional structure A(S) = (x 1 , y 1 , z 1 ), . . . , (x n , y n , z n ), we define a notion of biologically similar sequences V using Block Substitution Matrices (BLOSUM) (Henikoff and Henikoff 1992), and then employ formulations of adversarial attacks (Goodfellow, McDaniel, and Papernot 2018) on PFNNs within this space of similar sequences to identify a sequence S adv ∈ V that produces a maximally different three-dimensional structure A(S adv ).
- We then compute the RMSD and GDT between the structures for the original and adversarial inputs (A(S) and A(S adv )), and use these metrics as the robustness measure.
- If the RMSD (GDT) is small (high), the response of the PFNN is deemed robust; a large (small) RMSD (GDT) indicates that the predicted structure is not robust.

### BLOSUM Similarity Measures
- The Hamming distance is not a good measure of sequence similarity
- A scoring matrix called the PAM250 matrix is based on empirical analysis of 1572 mutations observed in 71 families of closely-related proteins that are 85% or more identical after they have been aligned
- The PAM1 model-based scoring matrix was obtained by normalizing the frequency of mutations to achieve a 99% identity between homologous proteins
- Henikoff and Henikoff (1992) developed a sequence similarity measure that counts replacement frequencies in conserved blocks across different proteins

## Approach
- The paper discusses the idea of adversarial examples, which are examples of proteins that have been mutated so that they produce structures that are different from the original protein.
- The paper discusses the use of BLOSUM matrices to identify neighborhoods of a given sequence that are biologically similar and hence expected to have similar 3D structures.
- The paper focuses on the state-of-the-art AlphaFold model, the winner of the 1st place in CASP2020.

### Sequence Similarity Measures

### Output Structural Measure
- The three dimensional structure of a sequence of residues is an ordered n-tuple of threedimensional co-ordinates
- The RMSD and GDT measures are used to compare the similarity of two protein structures
- The GDT score returns a value in [0, 1] where 1 indicates identical structures

### Adversarial Attacks on PFNNs
- Small carefully crafted changes in a few pixels of input images cause well-trained neural networks with otherwise high accuracy to consistently produce incorrect responses in domains such as computer vision.
- Given a neural network A mapping a sequence S of residues to a three-dimensional geometry A(S) describing the structure of the protein, we seek to obtain a sequence S such that the sequence similarity measure D seq (S, S ) between S and S is small and some structural distance measure D str (A(S), A(S )) is maximized.
- This can be achieved by Original and Adversarial Sequences.
- In our experiments, we set L = 20 and D str as the RMSD measure.
- Given the discrete nature of the input sequences, well-known methods for generating adversarial examples (e.g. gradient-based methods) fail to produce valid and accurate results.
- As such, we propose a solution based on a brute-force exploration in the space of biologically similar sequences that, given a sequence of interest S with n residues, can be defined as where X n is the set of all possible sequences over X of length n, D ham is the hamming distance, and H is a predefined threshold.
- For long sequences, the search space can be extensively large. Therefore, we select random samples from V L,H (S) and choose the sequence that returns the maximum value based on the RMSD measure.
- Our approach to generating adversarial sequences falls under the class of black-box attacks.
- This means that we only have access to the output of the network (Papernot et al. 2017).

## Complexity
- The PAA problem is NP-complete.
- The CLIQUE problem is reduced to an instance of the PAA problem.
- There is a clique of size k in G if and only if there is a feasible solution S in = S to the reduced PAA instance.

## Experimental Results
- For our experimental setup, we use the default settings of the latest version of AlphaFold 1.
- For further details about each step, we refer the reader to (Jumper et al. 2021a) and its supplementary information.
- We include results from using the high-accuracy full database configuration of the initial AlphaFold MSA step along with the less accurate (and faster) reduced database option.
- In order to compute the RMSD and GDT, we need to employ an alignment algorithm.
- In this paper, we use the built-in alignment PyMOL procedure without outlier rejections (Schrödinger and DeLano).
- The parameters of PyMOL alignment are selected using the default settings, which include an outlier rejection cutoff of 2, a maximum number of outlier rejection cycles of 5, and the use of the structural superposition step.
- We note that these outliers only impact the calculations of the RMSD.
- Our adversarial sequences are generated by randomly sampling 20 sequences from the set V L,H in (5) with H = 5 and L = 20.
- Then, we pick the sequence that returns the maximum value in RMSD structural distance.
- We use an adversarial sequence generation method in which the locations of the flipped residues are selected based on three categories: residues with confidence values near the (i) minimum confidence score (MIN. category), (ii) the average score (AVG. category), and (iii) the maximum confidence score (MAX. category).
- Results are presented in Table 2. We observe that, in general, selecting residues with low or high confidence scores is not related to the amount of the induced RMSD at the output. As such, in our method, the locations of the flipped residues are selected independent of the confidence scores.

### COVID-19 Case Studies
- We apply our adversarial approach to 111 publicly available COVID-19 protein sequences as of the time of this writing
- We observe that, independent of the predicted structure of the original sequence, a small change in the input sequence results in significant changes in the output structures
- The resulting structural distances (similarities) measured in Å (percentage) are given in terms of the RMSD (GDT-TS) in the fourth (sixth) column of Table 9 for the full database configuration
- Furthermore, we report the results using GDT-HA in the eighth column. The high similarity between the original and adversarial sequences is observed from the third column
- The similarity percentage is calculated as 100(n − D ham (S, S ))/n, where D ham (S, S ) ≤ H = 5. The complete results of all the considered proteins, including reduced AlphaFold configuration, are provided in the supplementary material
- As observed from the RMSD and GDT results in Table 9, small changes in the input sequence corresponding to only five residues cause AlphaFold to predict structures that are highly divergent from the predicted structure of the original sequence
- The last column in Table 9 reports the total execution time (in days) of running the 20 adversarial sequences that were randomly selected from the set V L,H , which is shown to scale with the sequence length
- We only select 20 samples given the long time incurred by AlphaFold to predict the output structure. Additionally, in Table 13, we report the average (deviation) prediction confidence results as for all the residues (designated with subscript 'all') and for the 5 altered residues (subscript 'diff'). The standard deviation is denoted σ. We ob-serve that, independent of the average prediction confidence, the RMSD between the original and adversarial predicted structures is always high. This is noted for both the full and reduced database configurations of AlphaFold. Moreover, we observe that AlphaFold predicts the adversarial structure with similar confidence values to the original sequence (e.g., see the 4th and 8th columns in both tables)
- The same observation holds for the entire sequence and for the altered residues (columns 6 and 10)
- In Tables 14 and 15 of the supplementary, we break down GDT scores between the structures of the original and perturbed sequences based on the prediction confidence scores of the original sequence. We use the regions (1 to 4) defined by AlphaFold. As observed w.r.t. all regions, GDT scores are, in general, low. For the considered dataset, the values presented in Table 5 gauge the overall robustness of AlphaFold to adversarial sequences.
- As indicated in the documentation of AlphaFold, for better accuracy, the full database configuration incurs a higher execution time compared to the reduced database configuration. The reported average values of the RMSD and GDT-TS measures are 14.78Å and 37.95%, respectively.

## Conclusion
- Protein folding neural networks (PFNNs) are often susceptible to adversarial attacks in the form of minor perturbations to the input protein sequence.
- In this paper, we have presented the first work in this direction by demonstrating that PFNNs are often susceptible to adversarial attacks in the form of minor perturbations to the input protein sequence.
- These perturbations can induce great changes in the predicted protein structure and the resulting lack of robustness precludes the adoption of such PFNNs in safety-critical applications.
- We have employed standard protein structural distance and similarity to measure the robustness of AlphaFold.
- While the perturbation methods employed in this paper were basic for the purposes of illustrating the lack of robustness of PFNNs, the results presented herein can be readily used as a baseline for future work on adversarial attacks on PFNNs and their robustness.
